### ISSUE TITLE
Feature Request: Fetch & Serve the file via this plugin

### ISSUE THREAD

**Issue by CMCDragonkai** [techknowlogick/caddy-s3browser#45]
I am using an AWS key which I can confirm is capable of downloading the object from S3. I do this with `aws s3` command.

When I run the s3 proxy, it is able to show me all the files in the bucket. So directory listing works. However if I go to download a file, I end up with `AccessDenied`.

```
<Error>
<Code>AccessDenied</Code>
<Message>Access Denied</Message>
<RequestId>F49146529F2A4EBB</RequestId>
<HostId>
/D5b4bvdqQWpqSxRK3si14PBnVpEHF6RnsmrQS8+PkKdBmPmxmmmu/fpypABpw58wdqHdO47Ri8=
</HostId>
</Error>
```

I have to make my bucket completely public for the download to work via the proxy. But surely the proxy is meant to work when the bucket is not public. That's the whole point of providing the key right? This proxy is downloading the file and proxying the file to the client.

What could be causing this access denied problem?

**Comment by CMCDragonkai** [comment]
I'm starting to think that caddy-s3browser doesn't actually proxy requests to the caddy server as authenticated requests to the S3 bucket.

Does that mean caddy-s3browser is basically just using the unauthenticated object URL to download the file? If that's the case, then obviously caddy-s3browser won't be able to proxy a file from s3 if that bucket is blocking all public access.

If so, is there a way to make use of https://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html so that the bucket can be private while the caddy server has authenticated access to the private objects?

The link describes a method for constructing the `Authorization` header for signed requests to S3. And that should be possible to do dynamically given that `caddy-s3browser` has our authentication credentials when we start the server.

**Comment by CMCDragonkai** [comment]
Another link: https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html

Also we might need this: https://docs.aws.amazon.com/sdk-for-go/api/aws/signer/v4/

**Comment by techknowlogick** [comment]
> I'm starting to think that caddy-s3browser doesn't actually proxy requests

That's correct, s3 browser just provides a file listing of whats available in the bucket, for the file proxying that is left up to the proxy directive from caddy, which I don't believe is AWS aware.

> If so, is there a way to make use of https://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html so that the bucket can be private while the caddy server has authenticated access to the private objects?

That may be possible, although it would require generating a direct link to S3 instead of proxying the request.

**Comment by ThinkChaos** [comment]
I have this working locally. I'll submit a PR with this (among other changes) soon!

**Comment by techknowlogick** [comment]
Closed with #60
